Partial Application Syntax: Stage 1

byara opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/master/CONTRIBUTING.md

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Fixes #9342 <!-- remove the (`) quotes to link the issues -->
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  |
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->

- I modified `parseExprListItem` and added a new condition that checks for `tt.comma` and another `boolean` called `allowPlaceholder`. Then I check for the plugin and add the node `Partial`
- I added a boolean argument `allowPlaceholder` in `parseCallExpressionArguments` and `parseExprListItem` to able to check if using `?` is allowed or not, so I can throw and error.
- I added 9 tests.

nicolo-ribaudo commented on packages/babel-generator/src/generators/expressions.js:
```diff
@@ -125,6 +125,10 @@ export function OptionalMemberExpression(node: Object) {
   }
 }
 
+export function Partial() {
```
This name doesn't match with the name defined in the parser. Also, could you add a generator test?

byara replied:
Sure, I'll add tests for the generator.
Do you mean in: `packages/babel-parser/src/types.js`
I see that I made a mistake there. I'll fix it 👍

byara replied:
Additionally, I suppose, I can move this function to `src/generators/types.js`? Because we are not considering `Partial` (or if we call it `ArgumentPlaceholder`) to be an expression.

nicolo-ribaudo replied:
Yeah good idea (I still prefer ArgumentPlaceholder)

byara commented on packages/babel-parser/ast/spec.md:
````diff
@@ -603,6 +605,16 @@ Any expression node. Since the left-hand side of an assignment may be any expres
 ## Super
 
+```js
+interface PartialExpression <: Node {
````
Sure, I was just trying to follow the Grammar section of the proposal: [Grammar](https://github.com/tc39/proposal-partial-application#grammar). But I also think, `ArgumentPlaceholder` shows the purpose better.

byara commented on packages/babel-parser/src/parser/expression.js:
```diff
@@ -807,6 +809,9 @@ export default class ExpressionParser extends LValParser {
 
+      case tt.question:
+        node = this.startNode();
+        this.raise(node.start, "Partial Application syntax is not allowed");
```
I wanted to be able to throw a message whenever the user is using an invalid syntax.
I added a boolean `allowPlaceholder` to `parseExprListItem` and `parseCallExpressionArguments` and thought, I could set it to false whenever we do not want to allow the usage of `?`.
Maybe I could throw the error message with another conditional here:

```javascript
} else if (this.match(tt.question)) {
      if (!allowPlaceholder){
        const node = this.startNode();
        this.raise(node.start, "Partial Application syntax is not allowed");
      }
      this.expectPlugin("partialApplication");
      const node = this.startNode();
      this.next();
      elt = this.finishNode(node, "ArgumentPlaceholder");
}
```

nicolo-ribaudo replied:
The default error would be `unexpected ?`. it's ok either way.

danielcaldas replied:
I like the `allowPlaceholder` idea from @byara , it can evolve into something more specific and make the users aware of when a partial application is allowed and when it's not, taking examples from https://github.com/tc39/proposal-partial-application#syntax we can tell the user we can point the user on the right direction instead of having him figure out why he's seeing an `unexpected ?` error.

byara commented:
- I renamed the node to `ArgumentPlaceholder` and added tests for the `babel-generator`.
- I'm not sure about the type of the node though. In `babel-types/src/validators/generated/index.js` and `babel-types/src/definitions/experimental.js` I have put the node under `UnaryLike` but I'm not sure about it. 
- After I made the change in `babel-types/src/validators/generated/index.js`, adding `ArgumentPlaceholder` under `UnaryLike`,  `make build`, creates [a few functions automatically](https://github.com/babel/babel/pull/9343/commits/65cdb981110d235f78dc021cf82a011326661351), not sure if that is normal.
- for now, I removed the Error message, but if we decide it is a good idea, I'll put it back.
(I'm also working on my own implementation of the plugin for this proposal)

nicolo-ribaudo commented:
> I renamed the node to ArgumentPlaceholder and added tests for the babel-generator.

:+1:

> I'm not sure about the type of the node though. In babel-types/src/validators/generated/index.js and babel-types/src/definitions/experimental.js I have put the node under UnaryLike but I'm not sure about it.

`?` is not an unary operator, I don't think that there should be any alias.

> After I made the change in babel-types/src/validators/generated/index.js, adding ArgumentPlaceholder under UnaryLike, make build, creates a few functions automatically, not sure if that is normal.

All the files inside `*/generated/*` are automatically created by the build process. You should only edit the files inside `definitions`.

byara commented:
> ? is not an unary operator, I don't think that there should be any alias.

removed the alias

> All the files inside */generated/* are automatically created by the build process. You should only edit the files inside definitions.

removed my changes from the `generated` folder.

nicolo-ribaudo commented on packages/babel-types/src/definitions/experimental.js:
```diff
@@ -10,6 +10,15 @@ import {
+  visitor: ["argument"],
+  fields: {
+    argument: {
+      validate: assertNodeType("ArgumentPlaceholder"),
```
This would mean that the AST is something like this:
```jsonc
{
  "type": "ArgumentPlaceholder",
  "argument": {
    "type": "ArgumentPlaceholder",
    // ...
  }
}
```

It should only be this:
```js
defineType("ArgumentPlaceholder", {});
```

nicolo-ribaudo commented on packages/babel-parser/src/parser/expression.js:
```diff
@@ -1940,6 +1943,11 @@ export default class ExpressionParser extends LValParser {
       if (refTrailingCommaPos && this.match(tt.comma)) {
         refTrailingCommaPos.start = this.state.start;
       }
+    } else if (allowPlaceholder && this.match(tt.question)) {
```
Wdyt about this logic, to have better errors?

```js
} else if (this.match(tt.question)) {
  this.expectPlugin("partialApplication");
  if (allowPlaceholder) {
    this.raise(this.state.start, "Unexpected argument placeholder");
  }
  // ...
```

byara replied:
I like this! It has a clear message and does not confuse the user. The user will understand that they should not use `ArgumentPlaceholder` there. I'll apply it in my next push.

nicolo-ribaudo commented on packages/babel-parser/ast/spec.md:
```diff
@@ -36,7 +37,7 @@ These are the core @babel/parser (babylon) AST node types.
     - [ForStatement](#forstatement)
     - [ForInStatement](#forinstatement)
-    - [ForOfStatement](#forofstatement)
+  - [ForOfStatement](#forofstatement)
```
Are these indent changes auto-generated? :thinking:

byara replied:
Yeah, I didn't change anything about the list, I just added my change in spec and the list was generated for me 🤔

danez replied:
I don't think this file is autogenerated, maybe it was one of your ide extensions?
Can you revert this line and the on above in `Changes`?

byara replied:
You are right. Apparently, `Markdown All in One` was doing auto formatting.

nicolo-ribaudo commented on packages/babel-types/src/definitions/experimental.js:
```diff
@@ -10,6 +10,8 @@ import {
   classMethodOrDeclareMethodCommon,
 } from "./es2015";
 
+defineType("ArgumentPlaceholder", {});
```
We should also update the `CallExpression` definition to accept placeholder arguments

byara replied:
@nicolo-ribaudo You mean in `/packages/babel-types/src/definitions/core.js` I have to add another assert node type like this:
```
defineType("CallExpression", {
  visitor: ["callee", "arguments", "typeParameters", "typeArguments"],
  builder: ["callee", "arguments"],
  aliases: ["Expression"],
  fields: {
    callee: {
      validate: assertNodeType("Expression"),
    },
    arguments: {
      validate: chain(
        assertValueType("array"),
        assertEach(
          assertNodeType("Expression", "SpreadElement", "JSXNamespacedName", "ArgumentPlaceholder"),
        ),
      ),
    },
...
```

nicolo-ribaudo commented on packages/babel-plugin-syntax-partial-application/package.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "@babel/plugin-syntax-partial-application",
+  "version": "7.2.0",
+  "description": "Allow parsing of optional properties",
```
This description needs to be updated

byara commented:
@nicolo-ribaudo I was looking through the changes and I noticed that I have missed a few things:
- forbidding the usage with `SpreadElement `
- allowing the usage in `TemplateLiteral`
What do you think? Should I address them in this PR? or another one?

rbuckton commented:
Partial application in tagged templates was pretty much rejected in a previous TC39 meeting, so I plan to remove that from the proposal for the time being.

byara commented:
> > forbidding the usage with SpreadElement
> 
> Do you mean that `foo(?, ...arr)` isn't valid? Or are you talking about `foo(...)`?

Right now, we can parse `foo(?, ...arr)`. But as I was reading [this part](https://github.com/tc39/proposal-partial-application#support-for-restspread-) of the proposal, I had the impression that maybe we shouldn't parse it?

nicolo-ribaudo commented:
I think that that part is only disallowing the old proposed `foo(...)` syntax (rest placeholder), not `foo(?, ...a)`.

byara commented:
> I think that that part is only disallowing the old proposed `foo(...)` syntax (rest placeholder), not `foo(?, ...a)`.

I knew it got it wrong! Then I guess everything is alright!
I'm also ready to make a PR for the plugin (It's like 90% done).

nicolo-ribaudo commented:
Lets just wait for anather review so that this PR can be merged. Can you rebase it in the meantime?

byara commented:
> Lets just wait for anather review so that this PR can be merged. Can you rebase it in the meantime?

Sure! I'll do that 👍

danez commented on packages/babel-parser/ast/spec.md:
````diff
@@ -862,6 +865,14 @@ interface SpreadElement <: Node {
 }
 ```
 
+## ArgumentPlaceholder
````
on more level please `###`

danez commented on packages/babel-parser/ast/spec.md:
```diff
@@ -72,6 +73,7 @@ These are the core @babel/parser (babylon) AST node types.
     - [LogicalExpression](#logicalexpression)
       - [LogicalOperator](#logicaloperator)
     - [SpreadElement](#spreadelement)
+  - [ArgumentPlaceholder](#argumentplaceholder)
```
should also be intended one more level.

danez commented on packages/babel-parser/ast/spec.md:
```diff
@@ -601,6 +603,7 @@ interface Expression <: Node { }
 
 Any expression node. Since the left-hand side of an assignment may be any expression in general, an expression can also be a pattern.
 
+
```
not necessary

nicolo-ribaudo commented:
@byara Thank you! We will merge this PR when we decide to release v7.4.0. In the meantime, could you squash it and rebase the other PR on top of this one?

nicolo-ribaudo commented:
I'm merging this locally
